packages/state-transition/src/util/loadState.ts

```diff
@@ -0,0 +1,251 @@
+// this value rarely changes even after 3 months of data as monitored on mainnet in Sep 2023
+function loadInactivityScores(
+  migratedState: BeaconStateAltair,
+  seedState: BeaconStateAltair,
+  inactivityScoresBytes: Uint8Array
+): void {
+  // migratedState starts with the same inactivityScores to seed state
+  migratedState.inactivityScores = seedState.inactivityScores.clone();
+  const oldValidator = migratedState.inactivityScores.length;
+  // UintNum64 = 8 bytes
+  const newValidator = inactivityScoresBytes.length / 8;
+  const minValidator = Math.min(oldValidator, newValidator);
+  const oldInactivityScores = migratedState.inactivityScores.serialize();
+  const isMoreValidator = newValidator >= oldValidator;
+  const modifiedValidators: number[] = [];
+  findModifiedInactivityScores(
+    isMoreValidator ? oldInactivityScores : oldInactivityScores.subarray(0, minValidator * 8),
+    isMoreValidator ? inactivityScoresBytes.subarray(0, minValidator * 8) : inactivityScoresBytes,
+    modifiedValidators
+  );
+
+  for (const validatorIndex of modifiedValidators) {
+    migratedState.inactivityScores.set(
+      validatorIndex,
+      ssz.UintNum64.deserialize(inactivityScoresBytes.subarray(validatorIndex * 8, (validatorIndex + 1) * 8))
+    );
+  }
+
+  if (isMoreValidator) {
+    // add new inactivityScores
+    for (let validatorIndex = oldValidator; validatorIndex < newValidator; validatorIndex++) {
+      migratedState.inactivityScores.push(
+        ssz.UintNum64.deserialize(inactivityScoresBytes.subarray(validatorIndex * 8, (validatorIndex + 1) * 8))
+      );
+    }
+  } else {
+    if (newValidator - 1 < 0) {
+      migratedState.inactivityScores = ssz.altair.InactivityScores.defaultViewDU();
+    } else {
+      migratedState.inactivityScores = migratedState.inactivityScores.sliceTo(newValidator - 1);
+    }
+  }
+}
+
+function loadValidators(
+  migratedState: BeaconStateAllForks,
+  seedState: BeaconStateAllForks,
+  newValidatorsBytes: Uint8Array
+): number[] {
+  const seedValidatorCount = seedState.validators.length;
+  const newValidatorCount = Math.floor(newValidatorsBytes.length / VALIDATOR_BYTES_SIZE);
+  const isMoreValidator = newValidatorCount >= seedValidatorCount;
+  const minValidatorCount = Math.min(seedValidatorCount, newValidatorCount);
+  // migrated state starts with the same validators to seed state
+  migratedState.validators = seedState.validators.clone();
+  const seedValidatorsBytes = seedState.validators.serialize();
+  const modifiedValidators: number[] = [];
+  findModifiedValidators(
+    isMoreValidator ? seedValidatorsBytes : seedValidatorsBytes.subarray(0, minValidatorCount * VALIDATOR_BYTES_SIZE),
+    isMoreValidator ? newValidatorsBytes.subarray(0, minValidatorCount * VALIDATOR_BYTES_SIZE) : newValidatorsBytes,
+    modifiedValidators
+  );
+  for (const i of modifiedValidators) {
+    migratedState.validators.set(
+      i,
+      ssz.phase0.Validator.deserializeToViewDU(
```

Comment:
try to reuse the unchanged properties of the old validator

Reply:
At least re-use pubkey and withdrawal_credentials which take the most memory of all